tinygraph/tinygraph-rng.c

```diff
@@ -16,6 +22,184 @@
  *   Section 6.3.4 "PCG-RXS-M-XS"
  */
 
+#ifdef __AVX2__
+
+
+typedef struct tinygraph_rng {
+  TINYGRAPH_ALIGN(32) __m256i state;
+  TINYGRAPH_ALIGN(32) __m256i mul;
+  TINYGRAPH_ALIGN(32) __m256i inc;
+  TINYGRAPH_ALIGN(32) uint32_t cache[8];
+  uint32_t cache_idx;
+} tinygraph_rng;
+
+
+tinygraph_rng* tinygraph_rng_construct(void) {
+  tinygraph_rng *out = tinygraph_align_malloc(32, sizeof(tinygraph_rng));
+
+  if (!out) {
+    return NULL;
+  }
+
+  // Default state initializer, will always return
+  // the same sequence of pseudo-random numbers.
+
+  // We compute and store 8x 32bit random numbers
+  // in one go, the idx then keeps track of which
+  // to return to the user or when to re-compute
+
+  // The multiplier is fixed for all 8 rngs, but
+  // the increment is different (and must be odd)
+  // to select different streams of each rng. This
+  // allows us to e.g. seed each rng with the same
+  // user seed and still get distinct sequences.
+
+  *out = (tinygraph_rng) {
+    .state = _mm256_set1_epi32(UINT32_C(0x46b56677)),
+    .mul = _mm256_set1_epi32(UINT32_C(747796405)),
+    .inc = _mm256_set_epi32(
+        UINT32_C(2891336453), UINT32_C(2891336455), UINT32_C(2891336457), UINT32_C(2891336459),
+        UINT32_C(2891336461), UINT32_C(2891336463), UINT32_C(2891336465), UINT32_C(2891336467)),
```

Comment:
Instead of 8 consecutive odd numbers, should we sample 8 odd numbers from the full uint32 range as increment initializers here?